### PR TITLE
[49582] Build the modal dialog to share a work package

### PR DESCRIPTION
**Pull request by HDinger**
- [x] Add a feature flag
- [x] Add a button "Share"
  - [x] In full view
  - [x] In split view
- [x] On Click, a dialog opens
  - [x] User autocompleter
    - [x] Remove necessity to have `manage_members` to see all users.
    - [x] When adding an already added user, update that share's role and place it on top 
  - [x] Role selector
  - [x] Invite Button
  - [x] List of already invited users
- [x] Adding new members
  - [x] permission check
  - [x] Email sending
- [x] Deleting existing members
  - [x] permission check
- [x] Updating existing members
  - [x] permission check
- [x] Change in data (add or delete) leads to refresh of the modal content
- [x] Permission check (UI)
  - [x] Add prevented
  - [x] Update prevented
  - [x] Delete prevented 
- [x] The UI is build with Primer components
- [x] Build shared work package mailer view
- [x] Add Tests

### Out of scope

**Button**
* Show number of shares inside the button in the WP full view

**Modal Content**
* Checkboxes
* Filter for Type
* Filter for Role
* Show roles below user name

**Autocompleter**
* Show only users which are not invited yet
* Inline invite via autocompleter
* Show current role of a user in the autocompleter results
* Multiselect
* Share with groups

### Work packages

* https://community.openproject.org/wp/50176
* https://community.openproject.org/wp/50175
* https://community.openproject.org/wp/49582

### CODE REVIEW

File: app/views/sharing_mailer/shared_work_package.html.erb [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  <tr>
+    <%= placeholder_cell('20px', vertical: false) %>
+  </tr>
+</table>
+
+<table <%= placeholder_table_styles(width: '100%', style: 'width:100%;max-width:700px') %>>
+  <tr>
+    <%= placeholder_cell('12px', vertical: true) %>
+    <td>
+      <%= render partial: 'sharing_mailer/notification_mailer_header',
```
Comment (HDinger) [inline app/views/sharing_mailer/shared_work_package.html.erb]: Hi @aaron-contreras 

I'd like to avoid having to add all these extra templates (mailer_body, mailer_header and notification_row) as we have all of them already and its hard to maintain multiple ones.
I am aware that the current Figma mockups define the mails slightly different, so I talked to the product team again and they agreed, that the emails should look exactly like out existing ones. So there is no need to add additional templates anymore but you can just use the exisiting ones instead.

File: spec/features/work_packages/share_spec.rb [hunk 2]
```diff
@@ -125,10 +125,8 @@
       perform_enqueued_jobs
       expect(ActionMailer::Base.deliveries.size).to eq(1)
 
-      # Closing and reopening the modal will show the same state as before.
-      share_modal.close
-
-      share_modal.expect_closed
+      # Reopening the modal will show the same state as before.
+      work_package_page.visit!
```
Comment (aaron-contreras) [inline spec/features/work_packages/share_spec.rb:129]: 👍

File: app/components/work_packages/share/modal_body_component.html.erb [hunk 3]
```diff
@@ -0,0 +1,69 @@
+<%=
+  # Todo: doubled ID
+  component_wrapper do
+  # The identifier is a magic key also found in the wp-share modal template.
+    turbo_frame_tag wrapper_key do
```
Comment (oliverguenther) [inline app/components/work_packages/share/modal_body_component.html.erb]:
```suggestion
  # The identifier is a magic key also found in the wp-share modal template.
  component_wrapper('turbo-frame') do
```

component_wrapper is calling `wrapper_key` interall,y

File: frontend/src/app/features/work-packages/openproject-work-packages.module.ts [hunk 4]
```diff
@@ -463,6 +468,7 @@ import { RecentItemsService } from 'core-app/core/recent-items.service';
     WorkPackageSplitViewComponent,
     BackButtonComponent,
   ],
+  schemas: [CUSTOM_ELEMENTS_SCHEMA],
```
Comment (oliverguenther) [inline frontend/src/app/features/work-packages/openproject-work-packages.module.ts:471]: Why is this part necessary? I'm not seeing a opce- being used inside the new templates, so I don't think this would be required.

**Review by oliverguenther — CHANGES_REQUESTED** [review]
Some tiny code remarks

- [ ] Adding a new share doesn't seem to do anything

Kapture 2023-10-04 at 15 04 28

The turbo_stream response is made, but the list is not updated.

File: app/components/work_packages/share/modal_body_component.html.erb (hunk 3 above)
Comment (aaron-contreras) [reply app/components/work_packages/share/modal_body_component.html.erb]: I had addressed this in a separate branch with this one as the base. I cherry-picked the commit onto this branch: https://github.com/opf/openproject/pull/13752/commits/6874062bce4d13f945ed9772c7b1a31b86e09346

File: frontend/src/app/features/work-packages/openproject-work-packages.module.ts (hunk 4 above)
Comment (ulferts) [reply frontend/src/app/features/work-packages/openproject-work-packages.module.ts:471]: It was necessary back then because we had some back and forth with the autocompleter but you are right that it no longer is for the current solution so I removed it again.

Comment (ulferts) [reply frontend/src/app/features/work-packages/openproject-work-packages.module.ts:471]: Turns out I was wrong and we still need that declaration [for the `turbo-frame` tag](https://github.com/opf/openproject/actions/runs/6416153569/job/17419335271). But it does feel bad to disable the check for all custom tags within the module if all we want is angular to ignore one. I would be interested in knowing if there is a better way.

**Comment by ulferts** [comment]
I am merging this PR now. 

We can still address the issues, @oliverguenther mentioned later on.
